### PR TITLE
Add rsub.Scalar

### PR DESCRIPTION
**Pull request by gpetters94**
I feel like I'm doing something wrong with the amount of casts I have to do, but what I have is working. Please let me know if there's something I'm missing.

### CODE REVIEW

**Comment by gpetters94** [comment]
Not sure why this is failing - it's working fine on my end. Is there a way to get more verbose output from the build system?

**Comment by cathyzhyi** [comment]
This is what I got locally:
```
FAIL - "RsubModule_basic"

Unexpected outcome summary:

****** Failed tests - 1 tests
    FAIL - "RsubModule_basic"
        @ trace item #0 - call to "forward"
        @ output of call to "forward"
        ERROR: value (Tensor with shape=[3, 4] min=-9.458e+07, max=-0.08848, mean=-1.576e+07) is not close to golden value (Tensor with shape=[3, 4] min=-0.8964, max=-0.08848, mean=-0.471)


Summary:
    Failed: 1
```

**Comment by gpetters94** [comment]
That's really weird. Maybe it's a version thing - I'll try in a different environment.

**Comment by cathyzhyi** [comment]
the build always use the latest nightly pytorch. Can you try updating your pytorch version?

**Comment by gpetters94** [comment]
I thought I did, but let me try again.

**Comment by ramiro050** [comment]
This is a very sneaky bug! Something useful to remember is that when you get values that are overflowing in the result of your calculation, it is probably because you're doing calculations with a tensor that has not been initialized properly, which is what happened with my implementation of [sum](https://github.com/llvm/torch-mlir/pull/300#event-5275073659).

From looking at the `linalg.generic` op in the MLIR outputed by the unit test:

```mlir
#map = affine_map<(d0, d1) -> (d0, d1)>
module attributes {torch.debug_module_name = "MyModule"}  {
  func @forward(%arg0: tensor<?x?xf32>, %arg1: tensor<?xf32>) -> tensor<?x?xf32> {
    %c1 = arith.constant 1 : index
    %c0 = arith.constant 0 : index
    %c1_i64 = arith.constant 1 : i64
    %0 = tensor.dim %arg0, %c0 : tensor<?x?xf32>
    %1 = tensor.dim %arg0, %c1 : tensor<?x?xf32>
    %2 = linalg.init_tensor [%0, %1] : tensor<?x?xf32>
    %3 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%arg0 : tensor<?x?xf32>) outs(%2 : tensor<?x?xf32>) {
    ^bb0(%arg2: f32, %arg3: f32):  // no predecessors
      %4 = arith.sitofp %c1_i64 : i64 to f32
      %5 = arith.mulf %arg2, %4 : f32
      %6 = arith.subf %arg3, %5 : f32
      linalg.yield %6 : f32
    } -> tensor<?x?xf32>
    return %3 : tensor<?x?xf32>
  }
}
```

you can see that in the subtraction at the end, you're using `%arg3`, which is the output tensor for the `linalg.generic` op, which corresponds to `%2`. Since the tensor `%2` is initialized only in shape but not in value, the tensor is full of garbage values, which is leading to the overflow error that @cathyzhyi is seeing.

Now, the `linalg.generic` body should not be using the output tensor as part of the calculation. The reason this is happening, from what I understand, is because `payloadArgs` does not correspond exactly to the arguments of the function `rsub`. If you look here:

https://github.com/llvm/torch-mlir/blob/fd6a42b4dea375c866766225aa353c07c845c9da/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L1764-L1765

you see that `payloadArgs` is just the input tensors and the output tensor. Since there is only one tensor in the inputs to `rsub`, then `payloadArgs` looks like `{x, output_tensor}`. Since the adaptor assigns names to the vector of values you give it in the order that the names appear in the op definition, then `adaptor.other()` ends up being assigned to the output tensor. So when you do this,

https://github.com/llvm/torch-mlir/blob/fd6a42b4dea375c866766225aa353c07c845c9da/lib/Conversion/TorchToLinalg/TorchToLinalg.cpp#L1453

you're using the wrong value.

The way to solve this is to do the same thing you're doing with `alpha`. Create a value

```cpp
Value constant = promoteScalarToDtype(b, loc, operands[1], self.getType());
```

corresponding to the scalar input of `rsub`, and then use that in the subtraction:

```cpp
return b.create<arith::SubFOp>(loc, consant, mult);
```

This results in the following MLIR:

```mlir
map = affine_map<(d0, d1) -> (d0, d1)>
module attributes {torch.debug_module_name = "MyModule"}  {
  func @forward(%arg0: tensor<?x?xf32>, %arg1: tensor<?xf32>) -> tensor<?x?xf32> {
    %c1 = arith.constant 1 : index
    %c0 = arith.constant 0 : index
    %cst = arith.constant 0.000000e+00 : f64
    %c1_i64 = arith.constant 1 : i64
    %0 = tensor.dim %arg0, %c0 : tensor<?x?xf32>
    %1 = tensor.dim %arg0, %c1 : tensor<?x?xf32>
    %2 = linalg.init_tensor [%0, %1] : tensor<?x?xf32>
    %3 = linalg.generic {indexing_maps = [#map, #map], iterator_types = ["parallel", "parallel"]} ins(%arg0 : tensor<?x?xf32>) outs(%2 : tensor<?x?xf32>) {
    ^bb0(%arg2: f32, %arg3: f32):  // no predecessors
      %4 = arith.sitofp %c1_i64 : i64 to f32
      %5 = arith.truncf %cst : f64 to f32
      %6 = arith.mulf %arg2, %4 : f32
      %7 = arith.subf %5, %6 : f32
      linalg.yield %7 : f32
    } -> tensor<?x?xf32>
    return %3 : tensor<?x?xf32>
  }
}
```

Hopefully this fixes the issue!

File: e2e_testing/torchscript/elementwise.py [hunk 1]
```diff
@@ -310,3 +310,36 @@ def forward(self, x):
...
+    def __init__(self):
+        super().__init__()
+
+    @export
+    @annotate_args([
+        None,
+        ([-1, -1], torch.float32, True),
+    ])
+    def forward(self, x):
+        return torch.rsub(x, 0.0)
```
Comment (ramiro050) [inline e2e_testing/torchscript/elementwise.py:341]: nit: It might be better to use a different value other than `0.0` because zero can come from other places when they are initialized to default values. For example, In the bug that @cathyzhyi mentioned, the reason this was working on your machine was because the output tensor was being obtained from memory that had been initialized to all zeros, not because of the input being given, which can make things very confusing!

**Comment by gpetters94** [comment]
Okay, it looks like it's working now. You're a life saver, Ramiro.

**Review by cathyzhyi — APPROVED** [review]
